rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+Putting a comment block right before a function associates that documentation with that function,
+and `@something` indicates a roxygen2 command,
+so what this file is saying is:
+
+-   the function has two parameters called `input_file` and `output_file`;
+-   it returns the number of distinct words found; and
+-   we want to export it (i.e., we want it to be visible outside the package).
+
+Our function is now documented,
+but when we run `Check`,
+we still get a warning.
+After a bit more searching and experimentation,
+we discover that we need to load the `devtools` package in the console
+and run `devtools::document()` to regenerate documentation:
+it isn't done automatically.
+
+```{r eval=FALSE}
+library(devtools)
+devtools::document()
+```
+
+When we do this,
+we get good news and bad news:
+
+```
+Updating zipffreq documentation
+First time using roxygen2. Upgrading automatically...
+Updating roxygen version in /Users/gvwilson/merely/rse-r-package/zipffreq/DESCRIPTION
+Loading zipffreq
+Warning: The existing 'NAMESPACE' file was not generated by roxygen2, and will not be overwritten.
+Writing count_words.Rd
+```
+
+Ah---the tutorials did warn us about this.
+We need to delete `NAMESPACE` and re-run `devtools:document()` again:
+
+```
+Updating zipffreq documentation
+Writing NAMESPACE
+Loading zipffreq
+Writing NAMESPACE
+```
+
+Doing this creates this file:
+
+```{r, child="rse-r-package/zipffreq/NAMESPACE"}
+```
+
+The comment at the start tells roxygen2 it can overwrite the file,
+and reminds us that we shouldn't edit it by hand.
+The `export(count_words)` directive is what we really want:
+as you might guess from the name `export`,
+it tells the package builder to make this function visible outside the package.
+
+After doing this,
+we go into "Build...Configure build tools" and check "Generate documentation with Roxygen".
+Running `Check` again now gives us a clean bill of health.
+If we restart and install,
+we can now use `?count_words` in the console to view our help.
+
+### Exercise: Document a function
+
+FIXME: write this exercise once the CO2 manipulation functions exist.
+
+## What should I document? {#rse-r-package-docwhat}
+
+The answer to the question in this section's title depends on what stage of development you are in.
+If you are doing [exploratory programming](glossary.html#exploratory-programming),
+a short comment to remind yourself of each function's purpose is good enough.
+(In fact, it's probably better than what most people do.)
+That comment should begin with an active verb and describe
+how inputs are turned into outputs.
+If the function has any [side effects](glossary.html#side-effects),
+you should eliminate them.
+If you can't,
+you should describe them too.
+
+An active verb is something like "extract", "normalize", or "find".
+For example,
+these are all good one-line comments:
+
+-   "Create a list of current ages from a list of birth dates."
+-   "Clip signals to lie in [0...1]."
```

Comment:
This line I assume is Python-lingo?

Reply:
Signal processing rather than Python - amended.